File: frontend/src/stores/queue.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { api } from '../utils/api'

/**
 * Queue Store - Manages message queue state per session (Issue #500)
 */
export const useQueueStore = defineStore('queue', () => {
  // ========== STATE ==========

  // Per-session queue items: Map<sessionId, Array<QueueItem>>
  const queuesBySession = ref(new Map())

  // Per-session pause state: Map<sessionId, boolean>
  const pausedBySession = ref(new Map())

  // Per-session pagination state: Map<sessionId, { offset, hasMore, total, pendingCount }>
  const paginationBySession = ref(new Map())

  // ========== GETTERS ==========

  function getItems(sessionId) {
    return queuesBySession.value.get(sessionId) || []
  }

  function getPendingCount(sessionId) {
    return paginationBySession.value.get(sessionId)?.pendingCount
      ?? getItems(sessionId).filter(i => i.status === 'pending').length
  }

  function isPaused(sessionId) {
    return pausedBySession.value.get(sessionId) || false
  }

  function hasMore(sessionId) {
    return paginationBySession.value.get(sessionId)?.hasMore || false
  }

  // ========== ACTIONS ==========

  async function fetchQueue(sessionId) {
    try {
      const data = await api.get(`/api/sessions/${sessionId}/queue?limit=100&offset=0`)
      queuesBySession.value.set(sessionId, data.items || [])
      queuesBySession.value = new Map(queuesBySession.value)
      paginationBySession.value.set(sessionId, {
        offset: data.items?.length || 0,
        hasMore: data.has_more || false,
        total: data.total || 0,
        pendingCount: data.pending_count ?? null,
      })
      paginationBySession.value = new Map(paginationBySession.value)
      return data.items
    } catch (error) {
      console.error('Failed to fetch queue:', error)
      return []
    }
  }

  async function loadMore(sessionId) {
    const pagination = paginationBySession.value.get(sessionId)
    if (!pagination?.hasMore) return
    try {
      const { offset } = pagination
      const data = await api.get(`/api/sessions/${sessionId}/queue?limit=100&offset=${offset}`)
      const existing = queuesBySession.value.get(sessionId) || []
      const combined = [...existing, ...(data.items || [])]
      queuesBySession.value.set(sessionId, combined)
      queuesBySession.value = new Map(queuesBySession.value)
      paginationBySession.value.set(sessionId, {
        offset: offset + (data.items?.length || 0),
        hasMore: data.has_more || false,
        total: data.total || 0,
        pendingCount: data.pending_count ?? null,
      })
      paginationBySession.value = new Map(paginationBySession.value)
      return data.items
    } catch (error) {
      console.error('Failed to load more queue items:', error)
      return []
    }
  }

  async function enqueueMessage(sessionId, content, resetSession = null, metadata = null) {
    try {
      const body = { content }
      if (resetSession !== null) body.reset_session = resetSession
      if (metadata !== null) body.metadata = metadata

      const data = await api.post(`/api/sessions/${sessionId}/queue-message`, body)
      return data.item
    } catch (error) {
      console.error('Failed to enqueue message:', error)
      throw error
    }
  }

  async function cancelItem(sessionId, queueId) {
    try {
      const data = await api.delete(`/api/sessions/${sessionId}/queue/${queueId}`)
      return data.item
    } catch (error) {
      console.error('Failed to cancel queue item:', error)
      throw error
    }
  }

  async function requeueItem(sessionId, queueId) {
    try {
      const data = await api.post(`/api/sessions/${sessionId}/queue/${queueId}/requeue`)
      return data.item
    } catch (error) {
      console.error('Failed to requeue item:', error)
      throw error
    }
  }

  async function clearQueue(sessionId) {
    try {
      const data = await api.delete(`/api/sessions/${sessionId}/queue`)
      return data.cancelled_count
    } catch (error) {
      console.error('Failed to clear queue:', error)
      throw error
    }
  }

  async function pauseQueue(sessionId, paused) {
    try {
      const data = await api.put(`/api/sessions/${sessionId}/queue/pause`, { paused })
      pausedBySession.value.set(sessionId, paused)
      pausedBySession.value = new Map(pausedBySession.value)
      return data
    } catch (error) {
      console.error('Failed to pause/resume queue:', error)
      throw error
    }
  }

  /**
   * Handle queue_update WebSocket message.
   * Called from websocket store when a queue_update message arrives.
   */
  function handleQueueUpdate(sessionId, payload) {
    const action = payload.action

    // Re-fetch from page 1 to stay current with real-time changes
    fetchQueue(sessionId)

    // Update pause state from action
    if (action === 'paused') {
      pausedBySession.value.set(sessionId, true)
      pausedBySession.value = new Map(pausedBySession.value)
    } else if (action === 'resumed') {
      pausedBySession.value.set(sessionId, false)
      pausedBySession.value = new Map(pausedBySession.value)
    }
  }

  /**
   * Update pause state from session data (e.g., from state_change).
   */
  function updatePauseState(sessionId, paused) {
    pausedBySession.value.set(sessionId, paused)
    pausedBySession.value = new Map(pausedBySession.value)
  }

  return {
    // State
    queuesBySession,
    pausedBySession,
    paginationBySession,
    // Getters
    getItems,
    getPendingCount,
    isPaused,
    hasMore,
    // Actions
    fetchQueue,
    loadMore,
    enqueueMessage,
    cancelItem,
    requeueItem,
    clearQueue,
    pauseQueue,
    handleQueueUpdate,
    updatePauseState,
  }
})

File: src/routers/queue.py
"""Queue endpoints: /api/sessions/{session_id}/queue*"""

from fastapi import APIRouter, HTTPException, Request

from .. import relay_client
from ..exception_handlers import handle_exceptions
from ..session_backend import BackendMode


def build_router(webui) -> APIRouter:
    router = APIRouter()

    def _is_remote() -> bool:
        return webui.coordinator.backend_mode == BackendMode.REMOTE

    # Issue #498: queue state lives wherever the session actually runs — Pattern A
    # wholesale relay when REMOTE is configured. QueueProcessor.ensure_running()
    # gains a REMOTE guard (src/queue_processor.py) so the Hub's own local queue
    # loop never fires for sessions whose queue now lives on REMOTE.

    @router.post("/sessions/{session_id}/queue-message")
    @handle_exceptions("enqueue message", value_error_status=400)
    async def enqueue_message(session_id: str, request: Request):
        """Enqueue a message for a session."""
        if _is_remote():
            return await relay_client.forward(webui.coordinator, request)
        data = await request.json()
        content = data.get("content")
        if not content:
            raise HTTPException(status_code=400, detail="content is required")

        item = await webui.coordinator.enqueue_message(
            session_id=session_id,
            content=content,
            reset_session=data.get("reset_session"),
            metadata=data.get("metadata"),
        )

        return {
            "queue_id": item["queue_id"],
            "position": item["position"],
            "item": item,
        }

    @router.get("/sessions/{session_id}/queue")
    @handle_exceptions("get queue")
    async def get_queue(session_id: str, request: Request, limit: int = 100, offset: int = 0):
        """List queue items for a session, paginated."""
        if _is_remote():
            return await relay_client.forward(webui.coordinator, request)
        if not await webui.service.get_session_exists(session_id):
            raise HTTPException(status_code=404, detail="Session not found")
        result = await webui.coordinator.get_queue(session_id, limit=limit, offset=offset)
        return result

    @router.delete("/sessions/{session_id}/queue/history")
    @handle_exceptions("clear queue history")
    async def clear_queue_history(session_id: str, request: Request):
        """Remove terminal (sent/failed/cancelled) queue items."""
        if _is_remote():
            return await relay_client.forward(webui.coordinator, request)
        if not await webui.service.get_session_exists(session_id):
            raise HTTPException(status_code=404, detail="Session not found")
        count = await webui.coordinator.clear_queue_history(session_id)
        await webui._broadcast_queue_update(session_id, "history_cleared", {"count": count})
        return {"cleared_count": count}

    @router.delete("/sessions/{session_id}/queue/{queue_id}")
    @handle_exceptions("cancel queue item")
    async def cancel_queue_item(session_id: str, queue_id: str, request: Request):
        """Cancel a pending queue item."""
        if _is_remote():
            return await relay_client.forward(webui.coordinator, request)
        item = await webui.coordinator.cancel_queue_item(session_id, queue_id)
        if not item:
            raise HTTPException(status_code=404, detail="Queue item not found or not pending")

        await webui._broadcast_queue_update(session_id, "cancelled", item)
        return {"item": item}

    @router.post("/sessions/{session_id}/queue/{queue_id}/requeue")
    @handle_exceptions("requeue item")
    async def requeue_item(session_id: str, queue_id: str, request: Request):
        """Re-queue a sent or failed item at the front."""
        if _is_remote():
            return await relay_client.forward(webui.coordinator, request)
        item = await webui.coordinator.requeue_item(session_id, queue_id)
        if not item:
            raise HTTPException(status_code=404, detail="Queue item not found or not in sent/failed state")

        await webui._broadcast_queue_update(session_id, "enqueued", item)
        return {"item": item}

    @router.delete("/sessions/{session_id}/queue")
    @handle_exceptions("clear queue")
    async def clear_queue(session_id: str, request: Request):
        """Clear all pending queue items."""
        if _is_remote():
            return await relay_client.forward(webui.coordinator, request)
        if not await webui.service.get_session_exists(session_id):
            raise HTTPException(status_code=404, detail="Session not found")
        count = await webui.coordinator.clear_queue(session_id)
        await webui._broadcast_queue_update(session_id, "cleared", {"count": count})
        return {"cancelled_count": count}

    @router.put("/sessions/{session_id}/queue/pause")
    @handle_exceptions("pause queue")
    async def pause_queue(session_id: str, request: Request):
        """Pause or resume the queue."""
        if _is_remote():
            return await relay_client.forward(webui.coordinator, request)
        data = await request.json()
        paused = data.get("paused", True)
        success = await webui.coordinator.pause_queue(session_id, paused)
        if not success:
            raise HTTPException(status_code=400, detail="Failed to update queue pause state")

        action = "paused" if paused else "resumed"
        await webui._broadcast_queue_update(session_id, action, {"paused": paused})
        return {"paused": paused}

    @router.put("/sessions/{session_id}/queue/config")
    @handle_exceptions("update queue config")
    async def update_queue_config(session_id: str, request: Request):
        """Update queue configuration."""
        if _is_remote():
            return await relay_client.forward(webui.coordinator, request)
        data = await request.json()
        success = await webui.coordinator.update_queue_config(session_id, data)
        if not success:
            raise HTTPException(status_code=400, detail="Failed to update queue config")
        return {"config": data}

    return router
